### PR TITLE
[SC-84] Add liquidation and flashloan e2e tests

### PR DESCRIPTION
**Pull request by hexonaut**
Reopened from [here](https://github.com/marsfoundation/aave-helpers/pull/6) to rebase to the newest e2e overhaul.

### CODE REVIEW

File: src/ProtocolV3TestBase.sol [hunk 1]
```diff
@@ -392,6 +408,82 @@ contract ProtocolV3TestBase is CommonTestBase {
...
+    IPool pool,
+    address user,
+    address liquidator,
+    ReserveConfig memory collateralConfig,
+    ReserveConfig memory testAssetConfig,
+    uint256 amount
+  ) internal {
+    uint256 collateralAmountBefore = IERC20(collateralConfig.aToken).balanceOf(user);
+    this._borrow(testAssetConfig, pool, user, amount, false);
+    assertGt(IERC20(testAssetConfig.variableDebtToken).balanceOf(user), 0, 'DEBT_ZERO');
```
Comment (lucas-manuel) [inline src/ProtocolV3TestBase.sol]: This assertion isn't needed since its in the `_borrow` function already

File: src/ProtocolV3TestBase.sol [hunk 2]
```diff
@@ -392,6 +408,82 @@ contract ProtocolV3TestBase is CommonTestBase {
...
+    IPoolConfigurator configurator = IPoolConfigurator(
+      IPoolAddressesProvider(pool.ADDRESSES_PROVIDER()).getPoolConfigurator()
+    );
+    // set ltv/lt to 1bps which enables liquidation on the position
```
Comment (lucas-manuel) [inline src/ProtocolV3TestBase.sol]: Set

File: src/ProtocolV3TestBase.sol [hunk 3]
```diff
@@ -392,6 +408,82 @@ contract ProtocolV3TestBase is CommonTestBase {
...
+    vm.prank(IPoolAddressesProvider(pool.ADDRESSES_PROVIDER()).getACLAdmin());
+    configurator.configureReserveAsCollateral(
+      collateralConfig.underlying,
+      1,
+      1,
+      10001
+    );
+    _liquidate(collateralConfig, testAssetConfig, pool, liquidator, user, amount, true);
+    // hf should be low enough to have the entire position be liquidated
```
Comment (lucas-manuel) [inline src/ProtocolV3TestBase.sol]: HF

File: src/ProtocolV3TestBase.sol [hunk 4]
```diff
@@ -392,6 +408,82 @@ contract ProtocolV3TestBase is CommonTestBase {
...
+    assertEq(IERC20(testAssetConfig.variableDebtToken).balanceOf(user), 0, 'LIQUIDATION_DEBT_NOT_ZERO');
+    uint256 collateralAmountAfter = IERC20(collateralConfig.aToken).balanceOf(user);
+    assertLt(collateralAmountAfter, collateralAmountBefore, 'COLLATERAL_NO_DECREASE');
+    // within 1% error is fine
```
Comment (lucas-manuel) [inline src/ProtocolV3TestBase.sol]: Within

File: src/ProtocolV3TestBase.sol [hunk 5]
```diff
@@ -505,6 +597,24 @@ contract ProtocolV3TestBase is CommonTestBase {
...
+    ReserveConfig memory collateral,
+    ReserveConfig memory debt,
+    IPool pool,
+    address liquidator,
+    address user,
+    uint256 amount,
+    bool receiveAToken
+  ) internal {
+    deal2(debt.underlying, liquidator, amount);
+
```
Comment (lucas-manuel) [inline src/ProtocolV3TestBase.sol:609]: This function should include before/after assertions for:
```
collateral aToken balance of borrower

collateral balance of liquidator
collateral balance of aToken
collateral balance of treasury

debtToken balance of borrower

borrow balance of liquidator
borrow balance of aToken

File: src/ProtocolV3TestBase.sol [hunk 6]
```diff
@@ -392,6 +408,82 @@ contract ProtocolV3TestBase is CommonTestBase {
...
+    vm.prank(IPoolAddressesProvider(pool.ADDRESSES_PROVIDER()).getACLAdmin());
+    configurator.configureReserveAsCollateral(
+      collateralConfig.underlying,
+      1,
+      1,
+      10001
+    );
+    _liquidate(collateralConfig, testAssetConfig, pool, liquidator, user, amount, true);
+    // hf should be low enough to have the entire position be liquidated
+    assertEq(IERC20(testAssetConfig.variableDebtToken).balanceOf(user), 0, 'LIQUIDATION_DEBT_NOT_ZERO');
```
Comment (lucas-manuel) [inline src/ProtocolV3TestBase.sol:434]: Can move this assertion into `_liquidate`

File: src/ProtocolV3TestBase.sol [hunk 7]
```diff
@@ -392,6 +408,82 @@ contract ProtocolV3TestBase is CommonTestBase {
...
+    uint256 collateralAmountAfter = IERC20(collateralConfig.aToken).balanceOf(user);
+    assertLt(collateralAmountAfter, collateralAmountBefore, 'COLLATERAL_NO_DECREASE');
+    // within 1% error is fine
+    assertApproxEqRel(
+      IERC20(collateralConfig.aToken).balanceOf(liquidator),
+      collateralAmountBefore - collateralAmountAfter,
+      0.01e18,
+      'LIQUIDATOR_NO_RECEIVE_COLLATERAL'
+    );
```
Comment (lucas-manuel) [inline src/ProtocolV3TestBase.sol]: This difference will be the increase in the treasury based on the `liquidationProtocolFeeAmount`, all assertions can be reliably calculated exactly

File: src/ProtocolV3TestBase.sol [hunk 8]
```diff
@@ -392,6 +408,82 @@ contract ProtocolV3TestBase is CommonTestBase {
...
+  ) internal {
+    assertEq(IERC20(testAssetConfig.underlying).balanceOf(address(this)), 0, 'UNDERLYING_NOT_ZERO');
+    pool.flashLoanSimple(
+      address(this),
+      testAssetConfig.underlying,
+      amount,
+      abi.encode(address(pool)),
+      0
+    );
+    assertEq(IERC20(testAssetConfig.underlying).balanceOf(address(this)), 0, 'UNDERLYING_NOT_ZERO');
```
Comment (lucas-manuel) [inline src/ProtocolV3TestBase.sol:459]: I guess we can't really assert anything else here, all good - actually is there a fee for flashloans?

File: src/ProtocolV3TestBase.sol [hunk 9]
```diff
@@ -392,6 +408,82 @@ contract ProtocolV3TestBase is CommonTestBase {
...
+    return true;
+  }
+
+  function _e2eTestMintToTreasury(
+    IPool pool,
+    ReserveConfig memory testAssetConfig
+  ) internal {
+    address[] memory assets = new address[](1);
+    assets[0] = testAssetConfig.underlying;
+    pool.mintToTreasury(assets);
```
Comment (lucas-manuel) [inline src/ProtocolV3TestBase.sol:483]: Need to add all token state change assertions here

**Review by lucas-manuel — CHANGES_REQUESTED** [review]
Comments for more assertions

File: src/ProtocolV3TestBase.sol (hunk 1 above)
Comment (hexonaut) [reply src/ProtocolV3TestBase.sol]: fixed.

File: src/ProtocolV3TestBase.sol (hunk 2 above)
Comment (hexonaut) [reply src/ProtocolV3TestBase.sol]: done

File: src/ProtocolV3TestBase.sol (hunk 3 above)
Comment (hexonaut) [reply src/ProtocolV3TestBase.sol]: done

File: src/ProtocolV3TestBase.sol (hunk 4 above)
Comment (hexonaut) [reply src/ProtocolV3TestBase.sol]: done

File: src/ProtocolV3TestBase.sol (hunk 6 above)
Comment (hexonaut) [reply src/ProtocolV3TestBase.sol:434]: It actually depends on if the HF is below 95% or not, so for here it makes more sense to be outside imo.

File: src/ProtocolV3TestBase.sol (hunk 8 above)
Comment (hexonaut) [reply src/ProtocolV3TestBase.sol:459]: Yes there is an optional fee. Aave has it turned on, Spark doesn't.

File: src/ProtocolV3TestBase.sol (hunk 7 above)
Comment (hexonaut) [reply src/ProtocolV3TestBase.sol]: Went down to 1bps. It's not super important it's accurate down to the last decimal. Could be improved later on if needed.

**Review by lucas-manuel — APPROVED** [review]
Approving first pass, will add more assertions in new PR

**Pull request merged**